src/metadata.py

```diff
@@ -0,0 +1,89 @@
+import json
+import redis
+from abc import ABC, abstractmethod
+from typing import Any, Dict, List, Optional
+
+DEFAULT_CACHE_EXPIRE_TIME = 60 * 60 * 12
+
+Metadata = Dict[str, Any]
+
+
+class MetadataPlugin(ABC):
+    #: List of plugin identifiers that plugin depends on
+    __depends_on__: List[str] = []
+    #: Enables cache for extracted metadata
+    __cacheable__: bool = False
+    #: Overrides default cache expire time
+    __cache_expire_time__: int = DEFAULT_CACHE_EXPIRE_TIME
+
+    def __init__(self) -> None:
+        self.redis: Optional[redis.StrictRedis] = None
+
+    @property
+    def name(self) -> str:
+        return self.__class__.__name__
+
+    def set_redis(self, redis: redis.StrictRedis) -> None:
+        self.redis = redis
+
+    def __rs_key(self, cache_tag: str) -> str:
+        return f"cached:{self.name}:{cache_tag}"
+
+    def _cache_fetch(self, cache_tag: str) -> Metadata:
+        if not self.redis:
+            return {}
+        rs_key = self.__rs_key(cache_tag)
+        obj = self.redis.get(rs_key)
```

Comment:
Not for this PR, but redis queries can be costly, maybe we could use cachetools to speed it up later :thinking:

Reply:
Yup, maybe it's better to keep it locally.